File: system-image-loader/src/main/java/com/levnovikov/system_image_loader/ImageLoaderImpl.java
package com.levnovikov.system_image_loader;

import android.graphics.BitmapFactory;
import android.util.Log;
import android.widget.ImageView;

import com.levnovikov.core_network.HttpClient;
import com.levnovikov.core_network.request.Request;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;

/**
 * Author: lev.novikov
 * Date: 27/3/18.
 */

public class ImageLoaderImpl implements ImageLoader {

    private HttpClient client;
    private String baseUrl;
    private File cache;

    private ImageLoaderImpl() {
    }

    @Override
    public void loadImage(ImageView imageView, String path) {
        saveImageInInternalStorage(path)
                .map(img -> BitmapFactory.decodeFile(img.getAbsolutePath()))
                .observeOn(AndroidSchedulers.mainThread())
                .map(bitmap -> {
                    imageView.setImageBitmap(bitmap);
                    return bitmap;
                })
                .subscribe(img -> {
                    Log.i(">>>IMAGE", "Image " + path + "is loaded");
                }, e -> {
                    e.printStackTrace();
                    throw new RuntimeException();
                });
    }

    private Single<File> saveImageInInternalStorage(String path) {
        Request request = new Request.Builder()
                .setMethod(Request.Method.GET)
                .setUrl(baseUrl + path)
                .build();
        return client.makeCall(request)
                .map(response -> {
                    File img = new File(cache, path.replace('/', '_'));
                    if (!img.exists()) {
                        Files.copy(response.getBody().getContentStream(), img.toPath(), StandardCopyOption.REPLACE_EXISTING);
                    }
                    return img;
                });
    }

    public static class Builder {
        private ImageLoaderImpl loader = new ImageLoaderImpl();

        public Builder setClient(HttpClient client) {
            loader.client = client;
            return this;
        }

        public Builder setBaseUrl(String baseUrl) {
            loader.baseUrl = baseUrl;
            return this;
        }

        public Builder setCache(File dir) {
            if (!dir.isDirectory()) throw new UnsupportedOperationException("Cache should be a directory");
            loader.cache = dir;
            return this;
        }

        public ImageLoader build() {
            validateLoader();
            return loader;
        }

        private void validateLoader() {
            if (loader.cache == null) throw new UnsupportedOperationException("Cache should be specified");
            if (loader.client == null) throw new UnsupportedOperationException("HttpClient should be specified");
            if (loader.baseUrl == null) throw new UnsupportedOperationException("Url should be specified");
        }
    }

}

File: system-image-loader/src/main/java/com/levnovikov/system_image_loader/ImageLoaderImpl.kt
package com.levnovikov.system_image_loader

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import android.util.LruCache
import com.levnovikov.core_network.HttpClient
import com.levnovikov.core_network.request.Request
import java.io.BufferedInputStream
import java.io.File
import java.io.FileOutputStream
import java.io.InputStream
import java.io.OutputStream
import java.nio.file.Files
import java.nio.file.StandardCopyOption

/**
 * Author: lev.novikov
 * Date: 27/3/18.
 */

private const val DEFAULT_IN_MEMORY_CACHE_SIZE = 200

class ImageLoaderImpl private constructor() : ImageLoader {

    private lateinit var client: HttpClient
    private lateinit var baseUrl: String
    private lateinit var persistentCache: File

    private var isTerminated = false

    private var inMemoryCacheSize = DEFAULT_IN_MEMORY_CACHE_SIZE
    private val inMemoryCache = LruCache<String, Bitmap>(inMemoryCacheSize)

    override fun loadImage(path: String, farm: Int): Bitmap? {
        if (isTerminated) throw UnsupportedOperationException("Loading after termination")
        inMemoryCache.get(path)?.let { return it.apply { Log.d(">>>IMG", "Cached image used") } }

        val img = saveImageInPersistentStorage(farm, path)
        Log.i(">>>IMG", "Image " + path + "is loaded")
        Log.d(">>>IMG", "Loaded image used")
        return BitmapFactory.decodeFile(img.absolutePath).apply { inMemoryCache.put(path, this) }
    }

    private fun saveImageInPersistentStorage(farm: Int, path: String): File {
        var img: File? = null
        var stream: InputStream? = null
        try {
            val request = Request.Builder()
                    .setMethod(Request.Method.GET)
                    .setUrl("http://farm$farm.$baseUrl${path}_s.jpg")
                    .build()
            val response = client.makeCall(request)
            img = File(persistentCache, "$path/$farm".replace('/', '_'))
            if (!img.exists()) {
                stream = response.body.contentStream
                copy(stream, img)
            }
            return img
        } catch (e: Exception) {
            img?.delete()
            Log.d(">>>IMG_DELETE", "delete file")
            throw e
        } finally {
            stream?.close()
        }
    }

    private fun copy(inputStream: InputStream, file: File) {
        val out = FileOutputStream(file)
        try {
            val buf = ByteArray(1024)
            var length: Int = inputStream.read(buf)
            while(length > 0){
                out.write(buf, 0, length)
                length = inputStream.read(buf)
            }
        } finally {
            out.close()
            inputStream.close()
        }
    }

    class Builder {
        private val loader = ImageLoaderImpl()

        fun setClient(client: HttpClient): Builder {
            loader.client = client
            return this
        }

        fun setBaseUrl(baseUrl: String): Builder {
            loader.baseUrl = baseUrl
            return this
        }

        fun setCache(dir: File): Builder {
            if (!dir.isDirectory) throw UnsupportedOperationException("Cache should be a directory")
            loader.persistentCache = dir
            return this
        }

        fun setInMemoryCacheSize(countOfImages: Int): Builder { //TODO will be better to specify size in bytes
            loader.inMemoryCacheSize = countOfImages
            return this
        }

        fun build(): ImageLoader {
            return loader
        }
    }

}
